rfd/0095-okta-integration.md

```diff
@@ -0,0 +1,405 @@
+      type: authentik
+      config:
+        authentik-field1: value1
+        authentik-field2: value2
+    ...
+```
+
+This configuration must be configured statically through a configuration file or command line
+options to isolate potentially sensitive values from the main Teleport process. This will
+ensure that Teleport Cloud users do not have to provide us with sensitive information. Names
+must be unique.
+
+### User specific configuration for identity providers
+
+The user spec will be expanded with a new per identity provider configuration that
+administrators can take advantage of. The configs are expected to map to the unique `name`
+(not type) of each provider.
+
+```yaml
+kind: user
+version: v2
+metadata:
+  name: alice
+spec:
+  roles: ['devs']
+  traits:
+    logins: ['admin']
+    kubernetes_groups: ['edit']
+  provider_configs:
+    okta:prod:
+      okta-field1: value1
+    okta:dev:
+      okta-field1: value1
+    authentik:prod:
+      authentik-field1: value1
+```
+
+It is up to each identity provider to implement the parsing and interpretation of these configs.
+
+### Background synchronization
+
+The background synchronization process, which synchronizes all applications from the identity
+provider without access information, is expected to run every 10 minutes, as the list of
+applications and groups is not expected to change much. Additionally, the potential to have an
+adverse impact on the identity provider by overusing APIs is non-zero, so it's preferred to
+minimize the number of calls to identity provider APIs.
+
+```mermaid
+sequenceDiagram
+   participant IdPSvc as IdP Integration Service
+   participant IdP as Identity Provider
+   participant AppSvc as Application Access Service
+
+   loop Every 10 minutes
+      loop For each identity provider
+         IdPSvc->>IdP: Get list of applications
+         IdPSvc->>AppSvc: Synchronize applications in Teleport
+
+         IdPSvc->>IdP: Get list of groups
+
+         IdPSvc->>AppSvc: Synchronize groups in Teleport
+      end
+   end
+```
+
+Each provider will be iterated through and all applications will be retrieved and inserted into
+the Teleport database with a label matching the name of the provider. A new "external groups"
+object will be created to house external groups, and users will be able to belong to these
+groups.
+
+### User specific provider applications and groups
+
+Another Part of the identity provider integration service will be a user level synchronization
+service that runs every 2 minutes.  When a user logs in, if Teleport has been configured for
+external provider access, users will be this service, which will determine provider specific 
+applications and groups the user has access to. This information will be stored in our database
+and synchronized every 2 minutes. All logged in users will have this synchronization process
+run while they are logged in, but not while they are logged out.
```

Comment:
Can this be determined on demand instead somehow? Using our RBAC system. For example, in your Teleport role you should be able to define which Okta apps/groups user has access to. Teleport knows about all Okta apps/groups (which API key allows it) but shows only those that match your role labels. Basically, the same as for any other Teleport resource (server, app, database).

Reply:
I've worked to integrate this on-demand functionality into the new version of the RFD,